File: python/docsearch/tokens.py
"""Offline token estimation.

Deliberately dependency-free: ingest must never make a network call, which
rules out tiktoken's first-use BPE download. Every threshold in the chunker is
approximate ("~1,200 tokens"), so a calibrated estimate is sufficient.

The estimator counts word and punctuation atoms and scales by a subword factor.
Measured against a Latin-script technical corpus this tracks BPE counts within
~10%.

Error direction matters more than error size. Under-counting is silent and
compounding: a chunk whose tokens are under-counted never trips the
subdivision cap, so it lands oversized and every size-based check downstream
reads a figure in the wrong unit. Over-counting only yields chunks smaller
than they needed to be. Where this file guesses, it guesses high.
"""

import re

_ATOM = re.compile(r"\w+|[^\w\s]")

#: Scripts written without spaces, where a character is roughly one BPE token
#: on its own: Han and its extensions, kana, Hangul syllables, CJK punctuation.
#: Word-atom counting reads an entire run of these as a single atom, which
#: under-counts by roughly 9x on Japanese and 4x on Chinese.
_CJK = re.compile(
    "["
    "\u3000-\u303f"  # CJK punctuation
    "\u3040-\u30ff"  # hiragana and katakana
    "\u3400-\u4dbf"  # Han, extension A
    "\u4e00-\u9fff"  # Han
    "\uf900-\ufaff"  # Han, compatibility ideographs
    "\uac00-\ud7af"  # Hangul syllables
    "]"
)

#: Letters in scripts this estimator has no calibration for. Cyrillic and Greek
#: are known to split more aggressively than Latin, and Arabic, Hebrew,
#: Devanagari and Thai are simply unmeasured here. Rather than invent a factor
#: that cannot be validated offline, their presence is reported so a document
#: relying on one is visible instead of silently mis-sized.
_UNCALIBRATED = re.compile(
    "["
    "\u0370-\u03ff"  # Greek
    "\u0400-\u04ff"  # Cyrillic
    "\u0530-\u058f"  # Armenian
    "\u0590-\u05ff"  # Hebrew
    "\u0600-\u06ff"  # Arabic
    "\u0900-\u097f"  # Devanagari
    "\u0e00-\u0e7f"  # Thai
    "\u10a0-\u10ff"  # Georgian
    "]"
)

_LETTER = re.compile(r"[^\W\d_]")

#: Average BPE sub-token splits per whitespace/punctuation atom for English
#: technical prose. Raising this makes the chunker more conservative.
SUBWORD_FACTOR = 1.3

#: One CJK character is about one token. Expressed relative to SUBWORD_FACTOR
#: so that count_atoms keeps returning a value tokens_from_atoms can scale,
#: preserving the accumulate-then-convert contract the chunker depends on.
_CJK_ATOM_WEIGHT = 1.0 / SUBWORD_FACTOR


def count_atoms(text: str) -> int:
    """Raw word/punctuation atoms in ``text``, before the subword factor.

    Accumulate these -- never summed :func:`estimate_tokens` results -- when
    measuring a sequence of pieces against a budget. ``estimate_tokens``
    truncates to int, so summing it over hundreds of short pieces compounds
    the rounding loss into hundreds of missing tokens and a budget check that
    silently never trips.
    """
    cjk = len(_CJK.findall(text))
    if not cjk:
        return len(_ATOM.findall(text))
    # Counted separately, then removed, so a run of ideographs is not also
    # counted once more as a single word atom.
    rest = len(_ATOM.findall(_CJK.sub(" ", text)))
    return rest + round(cjk * _CJK_ATOM_WEIGHT)


def tokens_from_atoms(atoms: int) -> int:
    """Convert an accumulated atom count to an estimated token count."""
    return int(atoms * SUBWORD_FACTOR)


def estimate_tokens(text: str) -> int:
    """Approximate the BPE token count of ``text``."""
    return tokens_from_atoms(count_atoms(text))


def uncalibrated_letter_share(text: str) -> float:
    """Share of letters written in a script this estimator cannot size.

    Latin and CJK are accounted for. Everything else -- Cyrillic, Greek,
    Arabic, Hebrew, Devanagari, Thai -- falls back to Latin word-atom counting,
    which is known to under-count them by a factor this project has no offline
    way to measure. A document made mostly of such text has chunk sizes, and
    therefore every size threshold applied to it, in an unknown unit.
    """
    letters = len(_LETTER.findall(text))
    if not letters:
        return 0.0
    return len(_UNCALIBRATED.findall(text)) / letters

File: internal/domain/tokens.go
package domain

// Offline token estimation.
//
// Deliberately dependency-free: ingest must never make a network call, which
// rules out downloading a tokenizer's vocabulary. Every threshold in the
// chunker is approximate ("~1,200 tokens"), so a calibrated estimate is
// sufficient. It counts word and punctuation atoms and scales by a subword
// factor, which tracks BPE within about 10% on Latin-script technical prose.
//
// Error direction matters more than error size. Under-counting is silent and
// compounding: a chunk whose tokens are under-counted never trips the
// subdivision cap, so it lands oversized and every size-based check downstream
// reads a figure in the wrong unit. Where this file guesses, it guesses high.

import (
	"math"
	"unicode"

	"github.com/bamsammich/docsearch/internal/pystr"
)

// SubwordFactor is the average number of BPE sub-tokens per whitespace or
// punctuation atom in English technical prose. Raising it makes the chunker
// more conservative.
const SubwordFactor = 1.3

// cjkAtomWeight makes one CJK character about one token, expressed relative
// to SubwordFactor so CountAtoms keeps returning a value TokensFromAtoms can
// scale.
const cjkAtomWeight = 1.0 / SubwordFactor

// _cjk covers scripts written without spaces, where a character is roughly
// one BPE token on its own. Word-atom counting reads a whole run of these as
// one atom, which under-counts by roughly 9x on Japanese and 4x on Chinese.
var _cjk = &unicode.RangeTable{
	R16: []unicode.Range16{
		{Lo: 0x3000, Hi: 0x303f, Stride: 1}, // CJK punctuation
		{Lo: 0x3040, Hi: 0x30ff, Stride: 1}, // hiragana and katakana
		{Lo: 0x3400, Hi: 0x4dbf, Stride: 1}, // Han, extension A
		{Lo: 0x4e00, Hi: 0x9fff, Stride: 1}, // Han
		{Lo: 0xac00, Hi: 0xd7af, Stride: 1}, // Hangul syllables
		{Lo: 0xf900, Hi: 0xfaff, Stride: 1}, // Han, compatibility ideographs
	},
}

// _uncalibrated covers scripts this estimator has no calibration for.
// Cyrillic and Greek split more aggressively than Latin, and the rest are
// unmeasured. Their presence is reported rather than guessed at.
var _uncalibrated = &unicode.RangeTable{
	R16: []unicode.Range16{
		{Lo: 0x0370, Hi: 0x03ff, Stride: 1}, // Greek
		{Lo: 0x0400, Hi: 0x04ff, Stride: 1}, // Cyrillic
		{Lo: 0x0530, Hi: 0x058f, Stride: 1}, // Armenian
		{Lo: 0x0590, Hi: 0x05ff, Stride: 1}, // Hebrew
		{Lo: 0x0600, Hi: 0x06ff, Stride: 1}, // Arabic
		{Lo: 0x0900, Hi: 0x097f, Stride: 1}, // Devanagari
		{Lo: 0x0e00, Hi: 0x0e7f, Stride: 1}, // Thai
		{Lo: 0x10a0, Hi: 0x10ff, Stride: 1}, // Georgian
	},
}

// CountAtoms returns the raw word and punctuation atoms in text, before the
// subword factor.
//
// Accumulate these, never summed EstimateTokens results, when measuring a
// sequence of pieces against a budget. EstimateTokens truncates, so summing
// it over hundreds of short pieces compounds the loss into a budget check that
// silently never trips.
//
// Python counts `\w+|[^\w\s]` matches after replacing CJK characters with
// spaces, then adds the CJK count scaled and rounded half to even. One scan
// does the same: a CJK character ends a word run and counts separately.
func CountAtoms(text string) int {
	atoms, cjk := 0, 0
	inWord := false
	for _, r := range text {
		switch {
		case unicode.Is(_cjk, r):
			cjk++
			inWord = false
		case pystr.IsWord(r):
			if !inWord {
				atoms++
				inWord = true
			}
		case pystr.IsSpace(r):
			inWord = false
		default:
			atoms++
			inWord = false
		}
	}
	if cjk == 0 {
		return atoms
	}
	return atoms + int(math.RoundToEven(float64(cjk)*cjkAtomWeight))
}

// TokensFromAtoms converts an accumulated atom count to an estimated token
// count, truncating toward zero as Python's int() does.
func TokensFromAtoms(atoms int) int {
	return int(float64(atoms) * SubwordFactor)
}

// EstimateTokens approximates the BPE token count of text.
func EstimateTokens(text string) int {
	return TokensFromAtoms(CountAtoms(text))
}

// UncalibratedLetterShare is the share of letters written in a script this
// estimator cannot size. A document made mostly of such text has chunk sizes,
// and every size threshold applied to it, in an unknown unit.
//
// As in Python, the numerator counts every character in those script blocks,
// letter or not, and the denominator counts letters.
func UncalibratedLetterShare(text string) float64 {
	letters, uncal := 0, 0
	for _, r := range text {
		if pystr.IsLetter(r) {
			letters++
		}
		if unicode.Is(_uncalibrated, r) {
			uncal++
		}
	}
	if letters == 0 {
		return 0
	}
	return float64(uncal) / float64(letters)
}
